### PR TITLE
Add spec to test using a tls connection

### PR DESCRIPTION
**Pull request by alieander**
Add tls_certs for both client and server, use the server certs when
spinning up nsqd's and the client certs for talking to it, helpers to
load the certs a little cleaner in the tests, and test the consumer and
producers ability to read and write when given an ssl_context.

nsq-cluster doesn't truely support tls yet meaning we can not turn on
things such as tls-required with it thinking nsqd's never start. This
will be useful in writing the negative tests in the future (ie: if
server tls, cannot write without it)

### CODE REVIEW

File: spec/support/tls_spec_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    {
+      key: tls_certs[:client_key],
+      certificate: tls_certs[:client_cert],
+      ca_certificate: tls_certs[:ca_cert]
+    }
+  end
+end
+
+RSpec.configure do |config|
+    config.include TlsSpecHelpers
```
Comment (thommahoney) [inline spec/support/tls_spec_helpers.rb]: spacing is tooooo biggggg

File: spec/support/tls_certs/nsq-client-cert.pem [hunk 2]
```diff
@@ -0,0 +1,17 @@
+-----BEGIN CERTIFICATE-----
+MIICpTCCAg6gAwIBAgIEMoanvTANBgkqhkiG9w0BAQsFADBtMQswCQYDVQQGEwJV
+UzEWMBQGA1UECBMNU2FuIEZyYW5jaXNjbzEJMAcGA1UEBxMAMREwDwYDVQQKEwhu
+c3EgcnVieTEUMBIGA1UEAxMLbnNxIHJ1YnkgQ0ExEjAQBgNVBAUTCTg0NzY4NTU2
+NDAeFw0xNjA0MTgyMjE0MzRaFw00OTEyMzEyMzU5NTlaMGsxCzAJBgNVBAYTAlVT
+MRYwFAYDVQQIEw1TYW4gRnJhbmNpc2NvMQkwBwYDVQQHEwAxETAPBgNVBAoTCG5z
+cSBydWJ5MRIwEAYDVQQDEwlsb2NhbGhvc3QxEjAQBgNVBAUTCTg0NzY4NTU2NTCB
+nzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA9JV79IJrbDnuDytYijCWVL8wCEYt
```
Comment (thommahoney) [inline spec/support/tls_certs/nsq-client-cert.pem]: this is supposed to be `nzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEA9JV79IJrbDnuDytYijCWVL8wCEYT`

File: spec/lib/nsq/tls_connection_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                                      ssl_context: ssl_context)
+  end
+
+  after do
+    @connection.close
+    @cluster.destroy
+  end
+
+  describe 'when using a tls connection' do
+    it 'can write a message onto the queue and read it back off again' do
```
Comment (thommahoney) [inline spec/lib/nsq/tls_connection_spec.rb]: is there some way of asserting that the message was actually sent using TLS? I see that you're passing `ssl_context` but I could also pass `foo: 'bar'` and nothing would happen differently.

File: spec/support/tls_certs/nsq-client-cert.pem (hunk 2 above)
Comment (alieander) [inline spec/support/tls_certs/nsq-client-cert.pem]: HAHAHAHAHAHAHAhahahahahaha.a.....a...a.....ha

File: spec/lib/nsq/tls_connection_spec.rb (hunk 3 above)
Comment (alieander) [inline spec/lib/nsq/tls_connection_spec.rb]: Unfortunately, there is not a simple solution to this...
As I described above, without changes in nsq-cluster to support tls in a more meaningful way (`tls-required` flag, `https` port definition and checking, etc.) this is harder than it should be to test.

Looking at the `msg` object that we are getting back here, it shows as having a connection that is using an instance of a `OpenSSL::SSL::SSLSocket` but you do not actually have access this socket to even check it's type without some ruby hackery.

The test that I will love to write in the future will involve changes to nsq-cluster and will stress the effect of `tls-required` and what happens when you attempt to read/write without tls.

File: spec/lib/nsq/tls_connection_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      topic = producer.topic
+      producer.write('some-tls-message')
+      wait_for { message_count(topic) == 1 }
+      expect(message_count(topic)).to eq(1)
+
+      consumer = new_consumer(ssl_context: ssl_context)
+      msg = consumer.pop
+      expect(msg.body).to eq('some-tls-message')
+      msg.finish
+
```
Comment (leklund) [inline spec/lib/nsq/tls_connection_spec.rb]: We were talking about this offline, but maybe we should add an assertion that the connection socket is a valid `OpenSSL::Socket` instance

File: spec/lib/nsq/tls_connection_spec.rb (hunk 3 above)
Comment (alieander) [inline spec/lib/nsq/tls_connection_spec.rb]: I added a check for the instance variable...

File: spec/lib/nsq/tls_connection_spec.rb (hunk 4 above)
Comment (alieander) [inline spec/lib/nsq/tls_connection_spec.rb]: did this here 92f4da8

**Comment by thommahoney** [comment]
👍

**Comment by leklund** [comment]
:+1:

**Pull request merged**